api/mysagw/oidc_auth/models.py

```diff
@@ -26,19 +26,42 @@ def is_staff(self):
 
 
 class OIDCUser(BaseUser):
+    def _validate_claims(self, claims):
+        for claim in [
+            settings.OIDC_ID_CLAIM,
+            settings.OIDC_EMAIL_CLAIM,
+            settings.OIDC_GROUPS_CLAIM,
+        ]:
+            if claim not in claims:
+                raise SuspiciousOperation(f'Couldn\'t find "{claim}" claim')
+
     def __init__(self, token: str, claims: dict):
         super().__init__()
+        self._validate_claims(claims)
 
         self.claims = claims
-        try:
-            self.username = self.claims[settings.OIDC_USERNAME_CLAIM]
-        except KeyError:
-            raise SuspiciousOperation("Couldn't find username claim")
-        self.groups = self.claims.get(settings.OIDC_GROUPS_CLAIM)
+        self.id = self.claims[settings.OIDC_ID_CLAIM]
+        self.username = self.claims[settings.OIDC_EMAIL_CLAIM]
+        self.groups = self.claims[settings.OIDC_GROUPS_CLAIM]
         self.group = self.groups[0] if self.groups else None
         self.token = token
         self.is_authenticated = True
-        self.identity, _ = Identity.objects.get_or_create(idp_id=self.username)
+
+        self.identity = None
+        for matcher in [{"idp_id": self.id}, {"email": self.username}]:
+            try:
+                self.identity = Identity.objects.get(**matcher)
+                self.identity.email = self.username
+                self.identity.idp_id = self.id
+            except Identity.DoesNotExist:
+                continue
+
+        if not self.identity:
+            self.identity = Identity.objects.create(
+                idp_id=self.id,
+                email=self.username,
+            )
+        self.identity.save()
```

Comment:
this line could be moved up to ca. line 55, to avoid a double save

Reply:
obsolete now.